packages/custom_lint/lib/src/v2/server_to_client_channel.dart

```diff
@@ -16,6 +17,146 @@ import '../plugin.dart';
 import 'custom_lint_analyzer_plugin.dart';
 import 'protocol.dart';
 
+/// A class that checks if there are conflicting packages in the context roots.
+@visibleForTesting
+class ConflictingPackagesChecker {
+  final _contextRootPackagesMap = <ContextRoot, List<Package>>{};
+  final _contextRootPubspecMap = <ContextRoot, Pubspec>{};
+
+  /// Adds a [contextRoot] and its [packages] to the checker.
+  /// We need to pass the [pubspec] to check if the package is a git dependency
+  /// and to check if the context root is a flutter package.
+  void addContextRoot(
+    ContextRoot contextRoot,
+    List<Package> packages,
+    Pubspec pubspec,
+  ) {
+    _contextRootPackagesMap[contextRoot] = packages;
+    _contextRootPubspecMap[contextRoot] = pubspec;
```

Comment:
I'm also getting the `pubspec` here so I can get better information from the git packages.

Unfortunately I couldn't find a way to the the information from transitive git dependencies, but I'm assuming that this would be a pretty rare edge case. But to be extra sure I've built a exceptional case for when this happen, the only thing is that I try extracting the git info from the package's root path.